src/main/java/org/jboss/remoting3/remote/OutboundMessage.java

```diff
@@ -104,36 +104,30 @@ public void accept(final Pooled<ByteBuffer> pooledBuffer, final boolean eof) thr
                 boolean timeoutExpired = false;
                 if (msgSize > 0 && ! sendCancel) {
                     // empty messages and cancellation both bypass the transmit window check
-                    for (;;) {
-                        final int currentWindow = window;
-                        if (window >= msgSize) {
-                            window -= msgSize;
-                            if (log.isTraceEnabled()) {
-                                log.tracef("Outbound message ID %04x: message window is open (%d-%d=%d remaining), proceeding with send", getActualId(), window + msgSize, msgSize, window);
-                            }
-                            break;
-                        }
-                        try {
-                            log.tracef("Outbound message ID %04x: message window is closed, waiting", getActualId());
-                            pipeOutputStream.wait(ackTimeout, 0);
-                            if (window == currentWindow) {
-                                // no changes, throw an exception
-                                timeoutExpired = true;
+                    if (!decrementWindow(msgSize)) {
+                        long ackTimeout = OutboundMessage.this.ackTimeout;
+                        long initialTime = System.currentTimeMillis();
+                        do {
+                            try {
```

Comment:
Also, you should always use `nanoTime` for timeouts, not wall clock time.

Reply:
@ropalka  thanks for reviewing! The intention is that exactly. This block is there just to protect against spurious wakes. I.e., if I want to wait for a specific time and I'm awaken before that timeout expires without a notify. So, by comparing the elapsed time since initialTime everytime I'm awaken and the condition is not met, I can certify that the timeout has really expired.
@dmlloyd  I was wondering that, I saw a PR to XNIO that used nanotime and converted that to ms, I'm going to do the same here.

Reply:
@ropalka  see my response to David below. Yes, you are correct, the initialization of intialTime is inconsistent with updating the field in the while loop.

Reply:
The problem is still there @fl4via , nanoTime() wasn't moved to do-while block.

Reply:
These timeout loops are a challenge every time. Part of the reason is that every blocking and timeout API works a little bit differently, so there's no single template that always works for all cases.

Because of this, my usual approach for these is to write the loop as being completely unwound, until I have enough repeated code that I can clearly see where the loop is and how it enters and exits, and then once I'm sure it's correct I can carefully refactor it into a cleaner loop "shape". This is something I learned after many, many mistakes.

Reply:
@ropalka the initialization of initialTime outside the block was a deliberate choice. I just forgot to add a final to that field, btw.
So, I'm initializing initialTime outside of the block. This marks the beginning of the time I'll wait. My expectations is that the normal path is without spurious wakes, so, I'll probably reach the end of the while block and leave the do/while after a single iteration. I.e., I'll either have been able to acquireWindow, or I'll have verified that the timeout is expired. In the event there was a spurious wake and the window was not acquired, I reenter the loop, after updating ackTimeout based on current time minus initialTime, a value that is decremented from the total ack timeout. So, ackTimeout field is updated and then I wait for the remaining amount of time I have available. Because I'm always recalculating ackTimeout based on the original timeout (i.e., OutboundMessage.this.ackTimeout), initialTime does not have to be updated at every iteration of the loop. Unless I'm missing something, in that case please point me where is the mistake and I'll fix it right away :-)

Reply:
Thanks for explanation @fl4via . Now I see what you had in mind.